superset/assets/src/query/Metric.ts

```diff
@@ -0,0 +1,98 @@
+import Column from './Column';
+import FormData from './FormData';
+
+export const LABEL_MAX_LENGTH = 43;
+
+export enum MetricKey {
+  METRIC = 'metric',
+  METRICS = 'metrics',
+  PERCENT_METRICS = 'percent_metrics',
+  RIGHT_AXIS_METRIC = 'metric_2',
+  SECONDARY_METRIC = 'secondary_metric',
+  X = 'x',
+  Y = 'y',
+  SIZE = 'size',
+}
+
+export enum Aggregate {
+  AVG = 'AVG',
+  COUNT = 'COUNT ',
+  COUNT_DISTINCT = 'COUNT_DISTINCT',
+  MAX = 'MAX',
+  MIN = 'MIN',
+  SUM = 'SUM',
+}
+
+export enum ExpressionType {
+  SIMPLE = 'SIMPLE',
+  SQL = 'SQL',
+}
+
+interface AdhocMetricSimple {
+  expressionType: ExpressionType.SIMPLE;
+  column: Column;
+  aggregate: Aggregate;
+}
+
+interface AdhocMetricSQL {
+  expressionType: ExpressionType.SQL;
+  sqlExpression: string;
+}
+
+export type AdhocMetric = {
+  label?: string,
+  optionName?: string,
+} & (AdhocMetricSimple | AdhocMetricSQL);
+
+type Metric = {
+  label: string;
+} & Partial<AdhocMetric>;
```

Comment:
oh weird, my comment here got lost.

Should this instead be a `| AdhocMetric ` instead of `& Partial<AdhocMetric>`? it seems like a `Metric` is _either_ a saved metric (with just a required label) or an `AdhocMetric` with required `expressionType` and either `sqlExpression` or `column`+`aggregate`. 

My making it a `Partial` aren't we losing the required `AdhocMetric` fields?

Reply:
Nope, `Partial` here makes `AdhocMetrics` optional in its entirety (we can test this by removing required fields like `aggregate` for the `Simple` `expressionType` columns in our unit tests and we are seeing type errors 😄). Since we are now requiring label field, this also makes `label` a required field in the resulting `Metric`s we return from `getMetrics()`.

We could alternatively write this as `{label: string} | (AdhocMetric & { label: string })`

Reply:
Gotcha,  `{label: string} | (AdhocMetric & { label: string })` makes sense based on my understanding.

> `Partial` here makes `AdhocMetrics` optional in its entirety

I thought that `Partial` made all fields optional, but if you do indeed get type errors upon removing `aggregate`, etc. I'm okay with this way of writing it 👍. Can you verify that and then we can merge (unless @kristw or @conglei have feedback)

Reply:
Yep. Verified that in the unit tests by removing `aggregate`. 
Btw, `Partial` does make all fields optional (it is what I meant by "optional in its entirety"), so your understanding on `Partial` is definitely correct. 😄